### PR TITLE
[clang][Sema] Warn consecutive builtin comparisons in an expression

### LINKED ISSUE #20456: Feature request: warn on `a < b < c`
|  |  |
| --- | --- |
| Bugzilla Link | [20082](https://llvm.org/bz20082) |
| Version | trunk |
| OS | All |

## Extended Description 
For C primers, it would be useful to warn on `a < b < c`: 

    int main(int argc, char* argv[])
    {
        int a = 2;
        int b = 3;
        int c = 4;
        if (a < b < c) {
            /* ... */
        }
    }

I think it makes sense to warn on these codes 
as the compiler emits warnings on codes like `if (a = b)`.

P.S. gcc emits warnings on such codes.

### PR DESCRIPTION
**Pull request by JOE1994**
Add warning under `-Wparentheses` for consistency with `gcc 14.1`.

Closes #20456

### CODE REVIEW

**Comment by AaronBallman** [comment]
> Set the warning to DefaultIgnore, although -Wparentheses is enabled by default 

Do we know why GCC elected to go that route? This seems like something that should be enabled by default because the false positive rate should be quite low.

**Comment by JOE1994** [comment]
> Do we know why GCC elected to go that route? This seems like something that should be enabled by default because the false positive rate should be quite low.

It seems like `gcc` doesn't enable `-Wparentheses` by default, while `clang` enables it by default.

I will add a follow-up commit to remove `DefaultIgnore` and update affected Clang tests :)

File: clang/test/Sema/parentheses.cpp [hunk 1]
```diff
@@ -215,3 +215,10 @@ namespace PR20735 {
     // fix-it:"{{.*}}":{[[@LINE-9]]:20-[[@LINE-9]]:20}:")"
   }
 }
+
+void consecutive_builtin_compare(int x, int y, int z) {
+  (void)(x < y < z);  // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x < y > z);  // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x < y <= z); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
+  (void)(x <= y > z); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
```
Comment (AaronBallman) [inline clang/test/Sema/parentheses.cpp:223]: I'd like an additional test along the lines of:
```
(void)((x < y) >= z);
```
and what about something along these lines?
```
struct S {
  bool operator<(const S &) const;
} s, t;

(void)(s < t >= z);
```
perhaps that should diagnose while this should not?
```
struct S {
  S operator<(const S &) const;
} a, b;

struct T {
  friend bool operator>=(const S &, const T &);
} c;

bool val = (a < b >= c);
```

Comment (JOE1994) [reply clang/test/Sema/parentheses.cpp:223]: (I'll refer to each of the 3 cases mentioned in the above comment as case1, case2, and case 3)

I pushed a commit that adds 2 `no-warning` test lines for case1.
It's unlikely for one to write `((x < y) >= z)` and believe it to have mathematical meaning of `(x < y) AND (y >= z)`.

I limited the scope of this PR to diagnosing consecutive "built-in" comparisons (as per the PR title).
Thus, the current implementation does not diagnose case 2.

I don't have the capacity to further work on making clang diagnose case 2 but not case 3.
I added a TODO note at the bottom of the test lines for now.

FYI, gcc doesn't diagnose any of case1, case2, or case 3.

Comment (AaronBallman) [reply clang/test/Sema/parentheses.cpp:223]: > I don't have the capacity to further make clang diagnose case 2 but not case 3.
> I added a TODO note at the bottom of the test lines for now.

SGTM!

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -487,6 +487,9 @@ Improvements to Clang's diagnostics
        }
      };
 
+- Clang emits a ``-Wparentheses`` warning for expressions with consecutive comparisons like ``x < y < z``.
+  It was made a ``-Wparentheses`` warning to be consistent with gcc.
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:491]:
```suggestion
- Clang emits a ``-Wparentheses`` warning for expressions with consecutive comparisons like ``x < y < z``.
  Fixes #GH20456.
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a small nit with the release notes.

File: clang/test/Sema/bool-compare.c [hunk 3]
```diff
@@ -85,7 +85,7 @@ void f(int x, int y, int z) {
   if ((a<y) != -1) {}// expected-warning {{comparison of constant -1 with boolean expression is always true}}
 
   if ((a<y) == z) {} // no warning
-  if (a>y<z)      {} // no warning
+  if (a>y<z)      {} // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
```
Comment (erichkeane) [inline clang/test/Sema/bool-compare.c:88]: I realize this is GCC's diagnostic, but I think it isn't particularly good of a diag.  I would prefer we:

1- At least match operators.  I REALLY wish we could print the actual expression (unless the expression complicated, at which point we print `<expr>` or something).  So something like: `comparison 'a>y<z' does not have its mathematical meaning`.

2- I REALLY wish we had a note after it with the range of the LHS here that says something like `Left comparison operator evaluates to `bool` before 'less-than' comparison with 'z'`.

Comment (Endilll) [reply clang/test/Sema/bool-compare.c:88]: Just throwing a wording idea: `chaining of comparison operators`.

File: clang/test/SemaTemplate/typo-template-name.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ namespace InExpr {
 
     // These are valid expressions.
     foo<foo; // expected-warning {{self-comparison}}
-    foo<int()>(0);
+    foo<int()>(0); // expected-warning {{comparisons like 'X<=Y<=Z' don't have their mathematical meaning}}
     foo<int(), true>(false);
```
Comment (shafik) [inline clang/test/SemaTemplate/typo-template-name.cpp:40]: It is a shame we don't catch this one but neither does gcc

**Pull request merged**